openbook_auth/models.py

```diff
@@ -1645,6 +1647,88 @@ def unmute_post_with_id(self, post_id):
         post = Post.objects.get(pk=post_id)
         return post
 
+    def create_invite(self, nickname):
+        self._check_can_create_invite(nickname)
+        UserInvite = get_user_invite_model()
+        invite = UserInvite.create_invite(nickname=nickname, invited_by=self)
+        self.invite_count = F('invite_count') - 1
+        self.save()
+        return invite
+
+    def update_invite(self, invite_id, nickname):
+        self._check_can_update_invite(invite_id)
+        UserInvite = get_user_invite_model()
+        invite = UserInvite.objects.get(id=invite_id)
+        invite.nickname = nickname
+        invite.save()
+        return invite
+
+    def get_user_invites(self, status_accepted=None, status_pending=None):
+        invites_query = Q(invited_by=self)
+        UserInvite = get_user_invite_model()
+
+        if status_accepted:
+            invites_query.add(Q(created_user__isnull=False), Q.AND)
+        elif status_pending:
+            invites_query.add(Q(created_user__isnull=True), Q.AND)
+
+        return UserInvite.objects.filter(invites_query)
+
+    def search_user_invites(self, query, status_accepted=None, status_pending=None):
+        invites_query = Q(invited_by=self, nickname__icontains=query)
+        UserInvite = get_user_invite_model()
+
+        if status_accepted:
+            invites_query.add(Q(created_user__isnull=False), Q.AND)
+        elif status_pending:
+            invites_query.add(Q(created_user__isnull=True), Q.AND)
+
+        return UserInvite.objects.filter(invites_query)
+
+    def delete_user_invite_with_id(self, invite_id):
+        self._check_can_delete_invite_with_id(invite_id)
+        UserInvite = get_user_invite_model()
+        invite = UserInvite.objects.get(id=invite_id)
+        self.invite_count = F('invite_count') + 1
+        self.save()
+        invite.delete()
+
+    def send_invite_to_invite_id_with_email(self, invite_id, email):
+        self._check_can_send_email_invite_to_invite_id(invite_id)
+        UserInvite = get_user_invite_model()
+        invite = UserInvite.objects.get(id=invite_id)
+        invite.email = email
+        invite.send_invite_email()
+
+    def _check_can_create_invite(self, nickname):
+        if self.invite_count == 0:
+            raise ValidationError(_('You have no invites left'))
+
+        UserInvite = get_user_invite_model()
+        if UserInvite.objects.filter(invited_by=self, nickname=nickname).exists():
+            raise ValidationError('Nickname already in use')
+
+    def _check_can_update_invite(self, invite_id):
+        self._check_is_creator_of_invite_with_id(invite_id)
+
+    def _check_can_send_email_invite_to_invite_id(self, invite_id):
+        self._check_is_creator_of_invite_with_id(invite_id)
```

Comment:
Spam opportunity? Perhaps we need to give the option once only. Otherwise through the API it could be used to send hundreds/thousands of emails on repeat.

Reply:
Yea I was thinking at first of checking too... but then many people's invites get lost or buried in a sea of emails.. and I can foresee people asking the user "Hey can you send me that invite again, I cant find it, or I dont use that email anymore, send it on this other email., etc." 

One option is to allow one email per email address, and if they want to send the invite to another email let them. One could technically still send the same invite to various multiple emails, but then again... if someone _really_ wants to do that they can always 1.create invite, send email, delete invite, repeat...  

Thoughts?

Reply:
Hmm perhaps we can just make it so that regardless of who you send the invite to you can only send it once every couple of hours or so.

Reply:
Oh wait, I saw the new code you added. I think that's fine. At least prevents spamming through the app. We need to get throttling in place for better protection.